fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PullUpPredicateRule.java

```diff
@@ -0,0 +1,166 @@
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.task.TaskContext;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class PullUpPredicateRule implements TreeRewriteRule {
+
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        RewriteContext rewriteContext = new RewriteContext();
+        root.getOp().accept(new Visitor(), root, rewriteContext);
+
+        //Convert all remaining predicates to LogicalFilter
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : rewriteContext.columnRefToConstant.entrySet()) {
+            OptExpression o = OptExpression.create(
+                    new LogicalFilterOperator(BinaryPredicateOperator.eq(entry.getKey(), entry.getValue())), root.inputAt(0));
+            root.setChild(0, o);
+        }
+        return root;
+    }
+
+    private static class Visitor extends OptExpressionVisitor<Void, RewriteContext> {
+
+        public Visitor() {
+        }
+
+        @Override
+        public Void visit(OptExpression optExpression, RewriteContext context) {
+            for (int childIdx = 0; childIdx < optExpression.getInputs().size(); ++childIdx) {
+                OptExpression child = optExpression.inputAt(childIdx);
+
+                RewriteContext rewriteContext = new RewriteContext();
+                child.getOp().accept(this, child, rewriteContext);
+                OptExpression c = handleLegacyPredicate(optExpression.getOutputColumns(), child, rewriteContext);
+                optExpression.setChild(childIdx, c);
+
+                context.columnRefToConstant.putAll(rewriteContext.columnRefToConstant);
+            }
+
+            return null;
+        }
+
+        public OptExpression handleLegacyPredicate(ColumnRefSet parentInputColumns, OptExpression optExpression, RewriteContext context) {
+            OptExpression root = optExpression;
+            if (parentInputColumns != null) {
+                Iterator<Map.Entry<ColumnRefOperator, ScalarOperator>> iterator =
+                        context.columnRefToConstant.entrySet().iterator();
+                while (iterator.hasNext()) {
+                    Map.Entry<ColumnRefOperator, ScalarOperator> entry = iterator.next();
+                    if (!parentInputColumns.contains(entry.getKey())) {
+
+                        // If the output columns does not contain this predicate,
+                        // leave the predicate under this node to provide more possibilities for subsequent predicate push down
+                        root = OptExpression.create(new LogicalFilterOperator(
+                                BinaryPredicateOperator.eq(entry.getKey(), entry.getValue())), root);
+                        iterator.remove();
+                    }
```

Comment:
If the context.columnRefToConstant has multiple entries, this seems to yeild a multiple levels filter. Can we zip the operators to one filter and add this filter to the top ?

Reply:
The way you mentioned is possible, but I don't think it is necessary. Because subsequent RULEs will optimize this pattern. I think no need to add complexity to the code here.